tempesta_fw/http_parser.c

```diff
@@ -4717,21 +4108,90 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
 static int
 __resp_parse_expires(TfwHttpMsg *msg, unsigned char *data, size_t len)
 {
-	int ret;
+	int r;
+	TfwHttpResp *resp = (TfwHttpResp *)msg;
+	TfwHttpParser *parser = &msg->conn->parser;
+
+	/*
+	 * A duplicate invalidates the header's value.
+	 * @resp->expires is set to zero - already expired.
+	 */
+	if (resp->cache_ctl.flags & TFW_HTTP_CC_HDR_EXPIRES)
+		parser->_i_st = I_ST(I_EoL);
+
+	r = __parse_http_date(msg, data, len);
+	if (r < CSTR_POSTPONE) {  /* (ret < 0) && (ret != POSTPONE) */
+		/*
+		 * On error just swallow the rest of the line.
+		 * @resp->expires is set to zero - already expired.
+		 */
+		parser->_date = 0;
+		parser->_i_st = I_ST(I_EoL);
+		r = __parse_http_date(msg, data, len);
+	}
+
+	if (r >= 0) {
+		resp->cache_ctl.expires = parser->_date;
+		resp->cache_ctl.flags |= TFW_HTTP_CC_HDR_EXPIRES;
+	}
+
+	return r;
+}
+
+static int
+__resp_parse_date(TfwHttpMsg *msg, unsigned char *data, size_t len)
+{
+	int r = CSTR_NEQ;
+	TfwHttpResp *resp = (TfwHttpResp *)msg;
+	TfwHttpParser *parser = &msg->conn->parser;
+
+	if (!test_bit(TFW_HTTP_B_HDR_DATE, resp->flags))
+		r = __parse_http_date(msg, data, len);
+
+	if (r < CSTR_POSTPONE) {  /* (ret < 0) && (ret != POSTPONE) */
+		/*
+		 * On error just swallow the rest of the line.
+		 * @resp->expires is set to zero - already expired.
+		 */
+		parser->_date = 0;
+		parser->_i_st = I_ST(I_EoL);
+		r = __parse_http_date(msg, data, len);
+	}
+
+	if (r >= 0) {
+		resp->date = parser->_date;
+		__set_bit(TFW_HTTP_B_HDR_DATE, resp->flags);
+	}
+
+	return r;
+}
+
+static int
+__resp_parse_if_modified(TfwHttpMsg *msg, unsigned char *data, size_t len)
+{
+	int r = CSTR_NEQ;
+	TfwHttpResp *resp = (TfwHttpResp *)msg;
 	TfwHttpParser *parser = &msg->conn->parser;
 
-	ret = __parse_http_date(msg, data, len);
-	if (ret < CSTR_POSTPONE) {  /* (ret < 0) && (ret != POSTPONE) */
+	if (!test_bit(TFW_HTTP_B_HDR_LMODIFIED, resp->flags))
+		r = __parse_http_date(msg, data, len);
+
+	if (r < CSTR_POSTPONE) {  /* (ret < 0) && (ret != POSTPONE) */
```

Comment:
By the way, GCC is smart enough to optimize `ret < 0 && ret != -1` to `ret < -1`.